### PR TITLE
Dispatch cache clear domain event

### PR DESCRIPTION
**Pull request by Prokyonn**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| License | MIT
| Documentation PR | sulu/sulu-docs# 

#### What's in this PR?

Dispatch domain event on `cacheClear`.

### CODE REVIEW

File: src/Sulu/Bundle/WebsiteBundle/EventSubscriber/CacheClearEventSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+use Doctrine\ORM\EntityManagerInterface;
+use Sulu\Bundle\ActivityBundle\Application\Collector\DomainEventCollectorInterface;
+use Sulu\Bundle\WebsiteBundle\Domain\Event\CacheClearedEvent;
+use Sulu\Bundle\WebsiteBundle\Event\CacheClearEvent;
+use Sulu\Bundle\WebsiteBundle\Events;
+use Sulu\Bundle\WebsiteBundle\ReferenceStore\WebspaceReferenceStore;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Adds the X-Generator header.
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/WebsiteBundle/EventSubscriber/CacheClearEventSubscriber.php]: ??

File: src/Sulu/Bundle/WebsiteBundle/EventSubscriber/CacheClearEventSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ */
+class CacheClearEventSubscriber implements EventSubscriberInterface
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/WebsiteBundle/EventSubscriber/CacheClearEventSubscriber.php]: would mark class as `@internal`

**Review by alexander-schranz — CHANGES_REQUESTED** [review]

**Review by alexander-schranz — APPROVED** [review]

File: src/Sulu/Bundle/WebsiteBundle/ReferenceStore/WebspaceReferenceStore.php [hunk 3]
```diff
@@ -23,6 +23,13 @@ public static function generateTagByWebspaceKey(string $webspaceKey): string
         return \sprintf('%s-%s', self::WEBSPACE_REFERENCE_ALIAS, $webspaceKey);
     }
 
+    public static function getWebspaceKeyFromTag(string $tag): string
+    {
+        $position = \strpos($tag, '-');
+
+        return \substr($tag, $position + 1);
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/WebsiteBundle/ReferenceStore/WebspaceReferenceStore.php]: I think I would go here with something like to throw an exception is it is not `webspace-...`:

```php
$parts = explode('-', $tag, 2);

if (!isset($parts[1]) || $parts[0] !== static::WEBSPACE_REFERENCE_ALIAS) {
      throw new \InvalidArgumentException('...');
}

return $parts[1];
```

Comment (niklasnatter) [inline src/Sulu/Bundle/WebsiteBundle/ReferenceStore/WebspaceReferenceStore.php]: think i would check if the tag starts with `WEBSPACE_REFERENCE_ALIAS + '-'` and only return a value if it does

File: src/Sulu/Bundle/WebsiteBundle/EventSubscriber/CacheClearEventSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+use Sulu\Bundle\WebsiteBundle\Domain\Event\CacheClearedEvent;
+use Sulu\Bundle\WebsiteBundle\Event\CacheClearEvent;
+use Sulu\Bundle\WebsiteBundle\Events;
+use Sulu\Bundle\WebsiteBundle\ReferenceStore\WebspaceReferenceStore;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * @internal
+ */
+class CacheClearEventSubscriber implements EventSubscriberInterface
```
Comment (niklasnatter) [inline src/Sulu/Bundle/WebsiteBundle/EventSubscriber/CacheClearEventSubscriber.php]: we called this `DomainEventSubscriber` in the `PageBundle`. think i would do that here too 🙂

File: src/Sulu/Bundle/WebsiteBundle/Domain/Event/CacheClearedEvent.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    /**
+     * @param mixed[]|null $payload
+     */
+    public function __construct(string $webspaceKey, ?array $payload)
+    {
+        parent::__construct();
+
+        $this->webspaceKey = $webspaceKey;
+        $this->payload = $payload ?? [];
```
Comment (niklasnatter) [inline src/Sulu/Bundle/WebsiteBundle/Domain/Event/CacheClearedEvent.php]: not sure if this a `payload`. i think i would store them in the `context['tags']`

File: src/Sulu/Bundle/WebsiteBundle/Domain/Event/CacheClearedEvent.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    public function getEventType(): string
+    {
+        return 'cache_cleared';
```
Comment (niklasnatter) [inline src/Sulu/Bundle/WebsiteBundle/Domain/Event/CacheClearedEvent.php:42]: can we add a translation for the event in this PR?

**Pull request merged**